_posts/2023-09-30-virtual-threads-4.adoc

```diff
@@ -0,0 +1,132 @@
+---
+layout: post
+title: 'Processing Kafka records on virtual threads'
+date: 2023-10-09
+tags: virtual-threads reactive redis kafka messaging
+synopsis: 'Learn about the virtual threads integration in Quarkus messaging (Kafka, AMQP, Pulsar...).'
+author: cescoffier
+---
+:imagesdir: /assets/images/posts/virtual-threads
+
+In https://quarkus.io/blog/virtual-threads-2/[another blog post], we have seen how you can implement a CRUD application with Quarkus to utilize virtual threads.
+The virtual threads support in Quarkus is not limited to REST and HTTP. 
+Many other parts support virtual threads, such as gRPC, scheduled tasks, and messaging.
+In this post, we will see how you can process Kafka records on virtual threads, increasing the concurrency of the processing.
+
+## Processing messages on virtual threads
+
+The Quarkus Reactive Messaging extension supports virtual thread.
+Similarly to HTTP, to execute the processing on a virtual thread, you need to use the `@RunOnVirtualThread` annotation:
+
+[source, java]
+----
+@Incoming("input-channel")
+@Outgoing("output-channel")
+@RunOnVirtualThread
+public Fraud detect(Transaction tx) {
+    // Run on a virtual thread
+}
+----
+
+The processing of each message runs on separate virtual threads. 
+So, for each message from the `input-channel`, a new virtual thread is created (as seen in https://quarkus.io/blog/virtual-thread-1/[this blog post], virtual thread creation is cheap).
+
+image::virtual-thread-messaging.png[Threading model of the messaging application,400,float="right",align="center"]
+
+This execution model can be used with any Quarkus reactive messaging connector, including AMQP 1.0, Apache Pulsar, and Apache Kafka.
+The concurrency of this processing is no longer limited by the number of worker threads, as it would with the `@Blocking` annotation.
+Thus, this novel execution model simplifies the development of highly concurrent data streaming applications. 
+
+As we will see later, such high-level concurrency can cause problems.
+To keep this concurrency controllable, Quarkus limits the number of concurrent message processing to `1024` (This default value is https://quarkus.io/guides/messaging-virtual-threads[configurable]).
+One of the main benefits of this limit is preventing the application from polling millions of messages, which would be very expensive in terms of memory.
+Without this limit, a Kafka application would poll all the records from the assigned topics-partitions and consume a large amount of memory.
+
+Also, you may wonder why we do not use virtual threads by default.
+The reasons have been explains in https://quarkus.io/blog/virtual-thread-1/#five-things-you-need-to-know-before-using-virtual-threads-for-everything[a previous blog post].
+Currently, there are limitations that can make virtual threads dangerous. 
```

Comment:
Also maybe add a link to the relevant section?

Reply:
It is the link from the previous sentence.

Reply:
Ah OK, there was a section at the bottom too but looks good!